runtime/major_gc.c

```diff
@@ -1077,3 +1088,89 @@ void caml_finalise_heap (void)
   while (caml_gc_phase == Phase_sweep)
     sweep_slice (LONG_MAX);
 }
+
+#if defined(NAKED_POINTERS_CHECKER) && defined(NATIVE_CODE)
+
+#ifdef _WIN32
+#define WIN32_LEAN_AND_MEAN
+#include <windows.h>
+
+Caml_inline int safe_load(volatile header_t * p, header_t * result)
+{
+  header_t v;
+  __try {
+    v = *p;
+  }
+  __except(GetExceptionCode() == EXCEPTION_ACCESS_VIOLATION ?
+        EXCEPTION_EXECUTE_HANDLER : EXCEPTION_CONTINUE_SEARCH) {
+    *result = 0xdeadbeef;
+    return 0;
+  }
+  *result = v;
+  return 1;
+}
+
+#else
+
+Caml_inline int safe_load (header_t * addr, /*out*/ header_t * contents)
+{
+  int ok;
+  header_t h;
+  intnat tmp;
+
+  asm volatile(
+      "leaq 1f(%%rip), %[tmp] \n\t"
+      "movq %[tmp], 0(%[handler]) \n\t"
+      "xorl %[ok], %[ok] \n\t"
+      "movq 0(%[addr]), %[h] \n\t"
+      "movl $1, %[ok] \n\t"
+  "1: \n\t"
+      "xorq %[tmp], %[tmp] \n\t"
+      "movq %[tmp], 0(%[handler])"
+      : [tmp] "=&r" (tmp), [ok] "=&r" (ok), [h] "=&r" (h)
+      : [addr] "r" (addr),
+        [handler] "r" (&(Caml_state->checking_pointer_pc)));
+  *contents = h;
+  return ok;
+}
+
+#endif
+
+static void check_pointer_safe (value v, value *p)
+{
+  header_t h;
+  tag_t t;
+
+  if (! safe_load(&Hd_val(v), &h)) goto on_segfault;
+
+  t = Tag_hd(h);
+  if (t == Infix_tag) {
+    v -= Infix_offset_hd(h);
+    if (! safe_load(&Hd_val(v), &h)) goto on_segfault;
+    t = Tag_hd(h);
+  }
+
+  /* For the pointer to be considered safe, either the given pointer is in heap
+   * or the (out of heap) pointer has a black header and its size is < 2 ** 40
+   * words (128 GB). If not, we report a warning. */
+  if (Is_in_heap (v) ||
```

Comment:
IIUC `check_pointer_safe` is only invoked when `!Is_in_heap(v) && !Is_young(v)`. `check_pointer_safe` should not be called on a pointer in the minor heap since OCaml does not define what the colour of minor objects is.

Would it be useful to rename this function to `caml_is_naked_pointer_safe`, with an assertion `CAMLassert (!Is_in_heap(v) && !Is_young(v))` to make these assumptions explicit?

Reply:
Yes, you're absolutely right!  I renamed to `is_naked_pointer_safe` (it's a static function, so no need for `caml_`), added the assertion, and removed the `Is_in_heap` test within.